### PR TITLE
feat: implemented a ScopeManager with a default scope

### LINKED ISSUE #345: Initialize scopes when creating hint context
Currently the unit testing suit creates the context to pass down to the hint execution without initializing any of its members. The idea for this is to only initialize them when necessary. However, most of the hints use the ScopeManager, and there should always be at least the main scope. Hence, let's modify the code so when the context is created, the Scope Manager is also initialized

### PR DESCRIPTION
**Pull request by fishonamos**
I Initialized DefaultContext function in the hinter package. This will initializes the ScopeManager with a default scope when creating a new HintRunnerContext. 

This ensures that the ScopeManager is always initialized with a default scope when necessary and resolve #345

### CODE REVIEW

**Review by TAdev0 — CHANGES_REQUESTED** [review]
@fishonamos hi, I think you just need to call `InitializeScopeManager` on `ctx` which is defined in scope.go file . You dont need to create a new `InitializeScopeManager` function.

Moreover, the first line of your function :
```
ctx.ScopeManager = *DefaultNewScopeManager()
```
already does the job of entering a new scope.  Indeed,  `DefaultNewScopeManager` function will execute `NewScopeManager(make(map[string]any)`, thus creating an  uninitialized scope which will be the main scope. The 2nd line will create another scope and it is not necessary.

Simply calling `ctx.InitializeScopeManager(make(map[string]any))` should work.

@cicr99  @rodrigo-pino  @MaksymMalicki  please correct me if I am wrong.

**Comment by fishonamos** [comment]
> @fishonamos hi, I think you just need to call `InitializeScopeManager` on `ctx` which is defined in scope.go file . You dont need to create a new `InitializeScopeManager` function.
> 
> Moreover, the first line of your function :
> 
> ```
> ctx.ScopeManager = *DefaultNewScopeManager()
> ```
> 
> already does the job of entering a new scope. Indeed, `DefaultNewScopeManager` function will execute `NewScopeManager(make(map[string]any)`, thus creating an uninitialized scope which will be the main scope. The 2nd line will create another scope and it is not necessary.
> 
> Simply calling `ctx.InitializeScopeManager(make(map[string]any))` should work.
> 
> @cicr99 @rodrigo-pino @MaksymMalicki please correct me if I am wrong.

That makes sense. Done it

File: pkg/hintrunner/hinter/hinter.go [hunk 1]
```diff
@@ -23,13 +23,19 @@ type HintRunnerContext struct {
 	ConstantSizeSegment mem.MemoryAddress
 }
 
+// InitializeScopeManager initializes the ScopeManager with a default scope.
+func (ctx *HintRunnerContext) InitializeScopeManager(scope map[string]any) {
+	ctx.ScopeManager = *DefaultNewScopeManager(scope)
+}
+
 func InitializeDefaultContext() *HintRunnerContext {
```
Comment (TAdev0) [inline pkg/hintrunner/hinter/hinter.go:31]: If i'm right with what i said before, no need to add a new function `InitializeScopeManager` here. 

Simply doing : 
```
func InitializeDefaultContext() *HintRunnerContext {
	ctx := &HintRunnerContext{
		DictionaryManager:         DictionaryManager{},
		SquashedDictionaryManager: SquashedDictionaryManager{},
		ConstantSizeSegment:       mem.UnknownAddress,
	}
	InitializeScopeManager(ctx, make(map[string]any)) // Initialize the ScopeManager with a default scope
	return ctx
}
```

should work

File: pkg/hintrunner/hinter/hinter.go [hunk 2]
```diff
@@ -24,12 +24,13 @@ type HintRunnerContext struct {
...
 
 func InitializeDefaultContext() *HintRunnerContext {
-	return &HintRunnerContext{
+	ctx := &HintRunnerContext{
 		DictionaryManager:         DictionaryManager{},
 		SquashedDictionaryManager: SquashedDictionaryManager{},
-		ScopeManager:              *DefaultNewScopeManager(),
 		ConstantSizeSegment:       mem.UnknownAddress,
 	}
+	InitializeScopeManager(ctx, make(map[string]any)) // Initialize the ScopeManager with a default scope
```
Comment (TAdev0) [inline pkg/hintrunner/hinter/hinter.go:32]: well, now i look at it again, it seems that the current main branch returns `&HintRunnerContext` with `ScopeManager: *DefaultNewScopeManager(),` . This is actually the same as doing  `InitializeScopeManager(ctx, make(map[string]any))` after removing the ScopeManager field. 

@cicr99  what do you think? I'm not sure this is what was asked with issue #345  . This file is the hinter.go and the hinter effectively creates a new context with a main scope uninitialized.

Comment (TAdev0) [reply pkg/hintrunner/hinter/hinter.go:32]: I guess the code that should be modified is in the `runHinterTests` function in zerohint_test.go  file? 

here : 

```
		ctx := &hinter.HintRunnerContext{}
		if tc.ctxInit != nil {
			tc.ctxInit(ctx)
		}
```
when running each test, the context is reset, with all its fields unitialized

**Comment by cicr99** [comment]
Hey @fishonamos 
For this issue what you need to modify is https://github.com/NethermindEth/cairo-vm-go/blob/b2602abd7a2901daba66ead1bc2805d200538ee2/pkg/hintrunner/zero/zerohint_test.go#L119 as @TAdev0 mentioned in his last comment
The idea is to initialize the scope manager for all tests and not having to do it manually.
In addition, you'll notice that there are some unit tests that have something like the following (see [here](https://github.com/NethermindEth/cairo-vm-go/blob/b2602abd7a2901daba66ead1bc2805d200538ee2/pkg/hintrunner/zero/zerohint_ec_test.go#L26C5-L28C7)):
```
ctxInit: func(ctx *hinter.HintRunnerContext) {
	ctx.ScopeManager.EnterScope(map[string]any{})
},
```
or (see [here](https://github.com/NethermindEth/cairo-vm-go/blob/b2602abd7a2901daba66ead1bc2805d200538ee2/pkg/hintrunner/zero/zerohint_signature_test.go#L123C5-L125C7)):
```
ctxInit: func(ctx *hinter.HintRunnerContext) {
	hinter.InitializeScopeManager(ctx, map[string]any{})
},
```
The idea for these, as you can see is just to initialize the ScopeManager, so once you modify the code to initialize it all the time, that won't be necessary, then those lines can be removed. Make sure to check all test cases for situations like these.

**Comment by TAdev0** [comment]
@fishonamos did you close the PR by mistake?

**Comment by fishonamos** [comment]
Just seeing this. I think it is a mistake.  But not an issue. I will raise
another today. I have found a way.

**Pull request closed without merging**